lib/keila/mailings/schemas/campaign.ex

```diff
@@ -47,6 +47,45 @@ defmodule Keila.Mailings.Campaign do
   def schedule_changeset(struct = %__MODULE__{}, params) do
     struct
     |> cast(params, [:scheduled_for])
-    |> validate_required([:scheduled_for])
+    |> validate_scheduled_for()
+  end
+
+  defp validate_scheduled_for(changeset) do
+    changeset
+    |> maybe_validate_old_scheduled_for()
+    |> validate_change(:scheduled_for, &maybe_validate_new_scheduled_for/2)
+  end
+
+  defp maybe_validate_old_scheduled_for(changeset = %{data: %__MODULE__{scheduled_for: nil}}),
+    do: changeset
+
+  defp maybe_validate_old_scheduled_for(changeset) do
+    scheduled_for = changeset.data.scheduled_for
+    {_offset, threshold} = campaign_schedule_offset_and_threshold()
+
+    case DateTime.compare(threshold, scheduled_for) do
+      :gt -> add_error(changeset, :scheduled_for, "is already about to be delivered")
+      _ -> changeset
+    end
+  end
+
+  defp maybe_validate_new_scheduled_for(:scheduled_for, nil), do: []
+
+  defp maybe_validate_new_scheduled_for(:scheduled_for, scheduled_for) do
+    {offset, threshold} = campaign_schedule_offset_and_threshold()
+
+    case DateTime.compare(threshold, scheduled_for) do
+      :gt -> [scheduled_for: "must be at least #{offset} seconds in the future"]
+      _ -> []
+    end
+  end
+
+  defp campaign_schedule_offset_and_threshold() do
+    offset =
+      Application.get_env(:keila, Keila.Mailings, []) |> Keyword.get(:campaign_schedule_offset, 0)
```

Comment:
Is `0` a sane default here? IMHO think we don't really want to fallback to anything here since we probably want to always have a offset unless we specify otherwise.

Reply:
Mhm … good question. Maybe a default is actually not even necessary at all since the five-minute default is set in `config.exs`. What do you think?

Reply:
I think we should rather raise to surface any configuration errors.

Reply:
Using Keyword.fetch! now to raise errors.